### PR TITLE
Time stretching 1 of 6 allow outward drag for stretch handles

### PR DESCRIPTION
**Pull request by saintmatthieu**
Resolves: #4850 

QA:
- [x] Test smart clip dragging (hidden data at ends)
- [x] Test that trim handles turn into stretch handles when pressing ALT

Small intentional behavior change:
- [x] dragging of smart clip boundaries allows 2 samples minimum, not 1 as before

*(short description of the changes and the motivation to make the changes)*

- [x] I signed [CLA](https://www.audacityteam.org/cla/)
- [x] The title of the pull request describes an issue it addresses
- [x] If changes are extensive, then there is a sequence of easily reviewable commits
- [x] Each commit's message describes its purpose and effects
- [x] There are no behavior changes unnecessary for the stated purpose of the PR

Recommended:
- [x] Each commit compiles and runs on my machine without known undesirable changes of behavior

### CODE REVIEW

File: libraries/lib-wave-track/WaveTrack.cpp [hunk 1]
```diff
@@ -2596,10 +2596,62 @@ void WaveTrack::GetEnvelopeValues(
...
+                   1e-9 ?
+                neighbour :
+                nullptr;
+}
+
+WaveClip*
+WaveTrack::GetAdjacentClip(const WaveClip& clip, PlaybackDirection direction)
+{
+   return const_cast<WaveClip*>(
+      const_cast<const WaveTrack&>(*this).GetAdjacentClip(clip, direction));
```
Comment (Paul-Licameli) [inline libraries/lib-wave-track/WaveTrack.cpp:2644]: Can use `as_const` once more

File: libraries/lib-wave-track/WaveTrack.h [hunk 2]
```diff
@@ -398,8 +399,36 @@ class WAVE_TRACK_API WaveTrack final
...
+   const WaveClip* GetAdjacentClip(
+      const WaveClip& clip, PlaybackDirection searchDirection) const;
+   /*!
+    * @copydoc GetAdjacentClip(const WaveClip&, PlaybackDirection) const
+    */
+   WaveClip*
+   GetAdjacentClip(const WaveClip& clip, PlaybackDirection searchDirection);
+
+   WaveClipConstHolders GetOverlappingClips(double t0, double t1) const;
+   WaveClipHolders GetOverlappingClips(double t0, double t1);
```
Comment (Paul-Licameli) [inline libraries/lib-wave-track/WaveTrack.h:430]: Declared but not defined in any of the six PRs.

File: src/tracks/playabletrack/wavetrack/ui/WaveClipAdjustBorderHandle.cpp [hunk 3]
```diff
@@ -75,19 +75,59 @@ namespace {
...
+   bool isStretchMode)
+{
+   if (!adjustingLeftBorder)
+      // A clip of one sample is invisible - who needs that ? Leave at least 2.
+      return clip.GetPlayStartTime() + 2.0 / clip.GetRate();
+
+   const auto prevClip = track.GetNextClip(clip, PlaybackDirection::backward);
+   if (isStretchMode)
+      return prevClip ? prevClip->GetPlayEndTime() :
+                        std::numeric_limits<double>::min();
```
Comment (Paul-Licameli) [inline src/tracks/playabletrack/wavetrack/ui/WaveClipAdjustBorderHandle.cpp:103]: Beware the difference of `min` and `lowest` for floating-point types!

**Review by Paul-Licameli — CHANGES_REQUESTED** [review]

File: src/tracks/playabletrack/wavetrack/ui/WaveClipAdjustBorderHandle.cpp [hunk 4]
```diff
@@ -75,19 +75,59 @@ namespace {
...
 WaveClipAdjustBorderHandle::AdjustPolicy::~AdjustPolicy() = default;
 
+namespace {
+double GetLeftAdjustLimit(
+   const WaveClip& clip, const WaveTrack& track, bool adjustingLeftBorder,
+   bool isStretchMode)
+{
+   if (!adjustingLeftBorder)
+      // A clip of one sample is invisible - who needs that ? Leave at least 2.
+      return clip.GetPlayStartTime() + 2.0 / clip.GetRate();
```
Comment (Paul-Licameli) [inline src/tracks/playabletrack/wavetrack/ui/WaveClipAdjustBorderHandle.cpp:98]: Was this little change from a 1 sample minimum to 2 asked for by anyone?

Are you sure it always works as intended?  Maybe somewhere there is some roundoff of the inexact sample time to a discrete sample count that goes off by one.

File: libraries/lib-wave-track/WaveTrack.h (hunk 2 above)
Comment (saintmatthieu) [reply libraries/lib-wave-track/WaveTrack.h:430]: Ok, then it's not used. I'll remove. Thanks.

File: src/tracks/playabletrack/wavetrack/ui/WaveClipAdjustBorderHandle.cpp (hunk 4 above)
Comment (saintmatthieu) [reply src/tracks/playabletrack/wavetrack/ui/WaveClipAdjustBorderHandle.cpp:98]: No, it wasn't, but in current release, if you trim so far that the visible part is only one sample and you release the trim handle, you cannot trim back (unless you Undo.) Let's ask @LWinterberg.

I think it should always work. The clip's start time is always displayed on a sample position.

File: src/tracks/playabletrack/wavetrack/ui/WaveClipAdjustBorderHandle.cpp (hunk 3 above)
Comment (saintmatthieu) [reply src/tracks/playabletrack/wavetrack/ui/WaveClipAdjustBorderHandle.cpp:103]: Great catch, thank you!

File: src/tracks/playabletrack/wavetrack/ui/WaveClipAdjustBorderHandle.cpp (hunk 4 above)
Comment (Paul-Licameli) [reply src/tracks/playabletrack/wavetrack/ui/WaveClipAdjustBorderHandle.cpp:98]: Please elaborate the comment I added at the top for the QA checklist.

**Review by Paul-Licameli — APPROVED** [review]

**Comment by dozzzzer** [comment]
Tested

**Pull request merged**